framework/src/modules/interoperability/mainchain/commands/cc_update.ts

```diff
@@ -12,18 +12,433 @@
+
+		const activeValidators: ActiveValidators[] = [];
+		const updatedValidators = [];
+		const keyList: Buffer[] = [];
+		const weights: bigint[] = [];
+		for (const v of validators) {
+			const { blsKey } = await this._validatorsAPI.getValidatorAccount(
+				context.getAPIContext(),
+				v.address,
+			);
+			updatedValidators.push({
+				bftWeight: v.bftWeight,
+				address: v.address,
+			});
+			activeValidators.push({
+				blsKey,
+				bftWeight: v.bftWeight,
+			});
+		}
+		sortValidatorsByBLSKey(activeValidators);
+		for (const v of activeValidators) {
+			keyList.push(v.blsKey);
+			weights.push(v.bftWeight);
+		}
+
+		const decodedCertificate = codec.decode<Certificate>(certificateSchema, txParams.certificate);
+
+		// if the CCU also contains a non-empty inboxUpdate, the certificate is only valid if it allows the sidechain account to remain live for a reasonable amount of time
+		if (
+			txParams.inboxUpdate &&
+			!(header.timestamp - decodedCertificate.timestamp < LIVENESS_LIMIT / 2)
+		) {
+			throw Error(
+				`Certificate is not valid as it passed Liveness limit of ${LIVENESS_LIMIT} seconds.`,
+			);
+		}
+		// Certificate and Validators Update Validity
+		const verifySignature = verifyWeightedAggSig(
+			keyList,
+			decodedCertificate.aggregationBits as Buffer,
+			decodedCertificate.signature as Buffer,
+			MESSAGE_TAG_CERTIFICATE.toString('hex'),
+			partnerChainAccount.networkID,
+			txParams.certificate,
+			weights,
+			certificateThreshold,
+		);
+
+		if (
+			!(decodedCertificate.height > partnerChainAccount.lastCertificate.height) ||
+			decodedCertificate.timestamp >= header.timestamp ||
+			!verifySignature
+		) {
+			throw Error(
+				`Certificate is invalid due to invalid last certified height or timestamp or signature.`,
+			);
+		}
+
+		const newActiveValidators = updateActiveValidators(
+			activeValidators,
+			txParams.activeValidatorsUpdate,
+		);
+		const validatorsHash = computeValidatorsHash(
+			newActiveValidators,
+			txParams.newCertificateThreshold || certificateThreshold,
+		);
+
+		if (!decodedCertificate.validatorsHash.equals(validatorsHash)) {
+			throw new Error('Validators hash is incorrect given in the certificate.');
+		}
+
+		// CCM execution
+		const interoperabilityStore = this.getInteroperabilityStore(context.getStore);
+		const decodedCCMs = txParams.inboxUpdate.crossChainMessages.map(ccm => ({
+			serialized: ccm,
+			deserilized: codec.decode<CCMsg>(ccmSchema, ccm),
+		}));
+		if (partnerChainAccount.status === CHAIN_REGISTERED) {
+			// If the first CCM in inboxUpdate is a registration CCM
+			if (
+				decodedCCMs[0].deserilized.crossChainCommandID === CROSS_CHAIN_COMMAND_ID_REGISTRATION &&
+				decodedCCMs[0].deserilized.receivingChainID === MAINCHAIN_ID
+			) {
+				partnerChainAccount.status = CHAIN_ACTIVE;
+			} else {
+				const beforeSendContext = createBeforeSendCCMsgAPIContext({
+					feeAddress: context.transaction.senderAddress,
```

Comment:
Out of this PR scope, but `feeAddress` was bit confusing during the token module implementation. maybe we could jest keep it as senderAddress.
Also, separate topic, token module needed ccm size, so we should add that as well